compiler/rustc_mir_transform/src/unreachable_prop.rs

```diff
@@ -57,42 +68,40 @@ impl MirPass<'_> for UnreachablePropagation {
 fn remove_successors<'tcx, F>(
     terminator_kind: &TerminatorKind<'tcx>,
-    predicate: F,
+    is_unreachable: F,
 ) -> Option<TerminatorKind<'tcx>>
 where
     F: Fn(BasicBlock) -> bool,
 {
-    let terminator = match *terminator_kind {
-        TerminatorKind::Goto { target } if predicate(target) => TerminatorKind::Unreachable,
-        TerminatorKind::SwitchInt { ref discr, switch_ty, ref targets } => {
+    let terminator = match terminator_kind {
+        // This will unconditionally run into an unreachable and is therefore unreachable as well.
+        TerminatorKind::Goto { target } if is_unreachable(*target) => TerminatorKind::Unreachable,
+        TerminatorKind::SwitchInt { targets, .. } => {
             let otherwise = targets.otherwise();
 
-            let original_targets_len = targets.iter().len() + 1;
-            let (mut values, mut targets): (Vec<_>, Vec<_>) =
-                targets.iter().filter(|(_, bb)| !predicate(*bb)).unzip();
+            let otherwise_is_unreachable = is_unreachable(otherwise);
 
-            if !predicate(otherwise) {
-                targets.push(otherwise);
-            } else {
-                values.pop();
-            }
-
-            let retained_targets_len = targets.len();
-
-            if targets.is_empty() {
+            // This case is equivalent to a goto terminator.
+            if targets.all_targets().len() == 0 && otherwise_is_unreachable
+            // If all targets are unreachable, we can be unreachable as well.
+                || otherwise_is_unreachable && targets.iter().all(|(_, bb)| is_unreachable(bb))
+            {
                 TerminatorKind::Unreachable
-            } else if targets.len() == 1 {
-                TerminatorKind::Goto { target: targets[0] }
-            } else if original_targets_len != retained_targets_len {
-                TerminatorKind::SwitchInt {
-                    discr: discr.clone(),
-                    switch_ty,
-                    targets: SwitchTargets::new(
-                        values.iter().copied().zip(targets.iter().copied()),
-                        *targets.last().unwrap(),
-                    ),
-                }
             } else {
+                // If there are multiple targets, don't delete unreachable branches. This looses information about
+                // reachability causing worse codegen.
+                // For example (see src/test/codegen/match-optimizes-away.rs)
+                //
+                // pub enum Two { A, B }
+                // pub fn identity(x: Two) -> Two {
+                //     match x {
+                //         Two::A => Two::A,
+                //         Two::B => Two::B,
+                //     }
+                // }
+                //
+                // This generates a `switchInt() -> [0: 0, 1: 1, otherwise: unreachable]`, which allows us or LLVM to
+                // turn it into just `x` later. Without the unreachable, such a transformation would be illegal.
```

Comment:
This statement is omly true because we don't have enough range information available, right?

Reply:
(Also: please add a codegen test showing this opt)

Reply:
`src/test/codegen/match-optimizes-away.rs` is the test for this, and found this issue in the opt originally. And this doesn't necessarily have to do with range information only, it could also be an `unreachable_unchecked` the user wrote themselves to enable another optimization. Deleting unreachable branches like this just isn't a good idea because it loses information.

Reply:
> This statement is only true because we don't have enough range information available, right?

Potentially, but IMO it's a bad idea to turn this:
```rust
match x {
    Enum::A => 1,
    Enum::B => 2,
    Enum::C => 3,
    
    // Not in the user's code, but added for switchInt's "otherwise" target:
    // _ => unreachable_unchecked(),
}
```
into:
```rust
match x {
    Enum::A => 1,
    Enum::B => 2,
    _ => 3,
}
```
It's erasing the refinement that the third case is only reachable when `x` is `Enum::C`.
You can recompute this refinement with an analysis, but why erase it in the first place?

See also https://github.com/rust-lang/rust/pull/99762#issuecomment-1195710504 where I suggest a solution (make `otherwise` optional).